src/shogun/mathematics/linalg/LinalgBackendViennacl.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2016, Shogun-Toolbox e.V. <shogun-team@shogun-toolbox.org>
+ * All rights reserved.
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Heiko Strathmann, Viktor Gal
+ */
+
+#include <shogun/mathematics/linalg/LinalgBackendBase.h>
+
+#ifndef Linalg_Backend_ViennaCL_H__
+#define Linalg_Backend_ViennaCL_H__
+
+#ifdef HAVE_VIENNACL
+
+#include <viennacl/vector.hpp>
+#include <viennacl/linalg/inner_prod.hpp>
+#include <shogun/mathematics/linalg/GPUMemoryViennaCL.h>
+
+namespace shogun
+{
+
+/** @brief linalg methods with ViennaCL backend */
+class LinalgBackendViennaCL : public LinalgBackendBase
+{
+	template <typename T>
+	friend struct GPUMemoryViennaCL;
+
+public:
+	/** vector dot-product macro */
+	#define BACKEND_GENERIC_DOT(Type) \
+	virtual Type dot(const SGVector<Type>& a, const SGVector<Type>& b) \
+	{  \
+		return dot_impl(a, b);  \
+	}
+
+	BACKEND_GENERIC_DOT(float32_t);
+	BACKEND_GENERIC_DOT(float64_t);
+	BACKEND_GENERIC_DOT(int32_t);
+
+	/** vector transfer from cpu to gpu method macro */
+	#define BACKEND_GENERIC_TO_GPU(Type) \
+	virtual GPUMemoryBase<Type>* to_gpu(const SGVector<Type>& vector) const \
+	{  \
+		return to_gpu_impl(vector);  \
+	}
+
+	BACKEND_GENERIC_TO_GPU(float32_t);
+	BACKEND_GENERIC_TO_GPU(float64_t);
+	BACKEND_GENERIC_TO_GPU(int32_t);
+
+	/** vector transfer from gpu to cpu method macro */
```

Comment:
I think these guys should also be doxygen documented with `@param` and `@return`